File: megpd/code/markov_megpd.r
library(evd)
library(tidyverse)
library(copula)

###########################################################
# 1. ROBUST CORE FUNCTIONS
###########################################################

dmegpd_biv <- function(x1, x2, kappa, sigma, xi, delta_func) {

  r <- x1 + x2

  # Radial components (unit scale)
  term_rad <- egpd::degpd_density(r,
    kappa = kappa, sigma = sigma,
    xi = xi
  )

  # Ensure delta is not too small to cause overflow
  d <- max(delta_func(r), 0.01)

  # Term 2: Angular part (Gaussian on log-ratios)
  log_ratio <- log(x1 / x2)
  # Stability check for the exponential
  z_sq <- (log_ratio / d)^2
  term_ang <- (1 / (sqrt(2 * pi) * d)) * exp(-0.5 * z_sq) * (r / (x1 * x2))

  term_rad * term_ang
}

##################################################
# Build grid
##################################################

make_grid <- function(x_prev, n_grid = 5000) {
  seq(1e-10, x_prev * 100, length.out = n_grid)
}

##################################################
# Evaluate conditional density
##################################################

conditional_pdf <- function(x_prev,
                            x_grid,
                            kappa,
                            sigma,
                            xi,
                            delta_func) {
  sapply(
    x_grid,
    function(x) {
      dmegpd_biv(
        x_prev,
        x,
        kappa,
        sigma,
        xi,
        delta_func
      )
    }
  )
}

##################################################
# Numerically integrate density -> CDF
##################################################

pdf_to_cdf <- function(x_grid,
                       pdf_vals) {
  dx <- diff(x_grid)

  # trapezoidal rule for CDF
  mid_pdf <- (
    pdf_vals[-1] +
      pdf_vals[-length(pdf_vals)]
  ) / 2

  cdf <- c(
    0,
    cumsum(mid_pdf * dx)
  )

  total_mass <- tail(cdf, 1)

  if (total_mass <= 0) {
    return(NULL)
  }

  cdf <- cdf / total_mass

  # numerical safeguard
  cdf <- cummax(cdf)

  list(
    cdf = cdf,
    total_mass = total_mass
  )
}

##################################################
# Inverse-CDF sampler
##################################################

sample_from_cdf <- function(x_grid,
                            cdf_vals,
                            u = runif(1)) {
  # approx(
  #   x = cdf_vals,
  #   y = x_grid,
  #   xout = u,
  #   ties = "ordered"
  # )$y

  keep <- c(TRUE, diff(cdf_vals) > 1e-12)

  F_inv <- splinefun(
    cdf_vals[keep],
    x_grid[keep],
    method = "monoH.FC"
  )
  F_inv(u)
}

###########################################################
# 2. UPDATED SAMPLER
###########################################################

##################################################
# Markov transition sampler
##################################################

markov_megpd_sampler <- function(x_prev,
                                 kappa,
                                 sigma,
                                 xi,
                                 delta_func,
                                 n_grid = 5000) {
  x_grid <- make_grid(
    x_prev,
    kappa = kappa,
    sigma = sigma,
    xi = xi,
    n_grid = n_grid
  )

  pdf_vals <- conditional_pdf(
    x_prev,
    x_grid,
    kappa,
    sigma,
    xi,
    delta_func
  )

  if (sum(pdf_vals) <= 0) {
    return(x_prev)
  }

  cdf_obj <- pdf_to_cdf(
    x_grid,
    pdf_vals
  )

  cat("\n Total integrated mass:", cdf_obj$total_mass, "\n")

  if (is.null(cdf_obj)) {
    print("Warning: CDF construction failed, returning previous value.")
    return(x_prev)
  }

  sample_from_cdf(
    x_grid,
    cdf_obj$cdf
  )
}

###########################################################
# 3. SIMULATION
###########################################################

###########################################################
# Markov chain simulation
###########################################################

simulate_megpd_chain <- function(
  n_steps = 2000,
  kappa,
  sigma,
  xi,
  delta_func,
  x0 = 1,
  burn_in_prop = 0.10,
  n_grid = 5000,
  show_progress = TRUE
) {
  if (burn_in_prop < 0 || burn_in_prop >= 1) {
    stop("burn_in_prop must be in [0, 1).")
  }

  x <- numeric(n_steps)
  x[1] <- x0

  if (show_progress) {
    pb <- txtProgressBar(min = 0, max = n_steps, style = 3)
  }

  for (t in 2:n_steps) {
    x[t] <- markov_megpd_sampler(
      x_prev = x[t - 1],
      kappa = kappa,
      sigma = sigma,
      xi = xi,
      delta_func = delta_func,
      n_grid = n_grid
    )

    if (show_progress) {
      setTxtProgressBar(pb, t)
    }
  }

  if (show_progress) {
    close(pb)
  }

  burn_in <- floor(n_steps * burn_in_prop)

  list(
    full_chain = x,
    final_chain = x[(burn_in + 1):n_steps],
    burn_in = burn_in
  )
}

set.seed(1)
n <- 2000
kappa_val <- 2
sigma_val <- 1
xi_val <- 0.5

# Dependence that gets STRONGER as values get LARGER
delta_strong_upper <- function(r) {
  # Starts at 0.8 (weak dependence at 0) and decays to 0.1 (strong dependence at infinity)
  0.2 + 0.6 * exp(-r / 5)
}
    
# Run
set.seed(1)

sim <- simulate_megpd_chain(
  n_steps = n,
  kappa = kappa_val,
  sigma = sigma_val,
  xi = xi_val,
  delta_func = delta_strong_upper,
  x0 = 1,
  burn_in_prop = 0.10,
  n_grid = 100
)

final_chain <- sim$final_chain

File: megpd/code/markov_megpd.jl
using QuadGK, Roots, ProgressMeter
using SpecialFunctions
using Distributions

function delta(r::Float64) 0.2 + 0.6*exp(-r/5) end
# function delta(r::Float64) 0.8 end

# gd = Gamma(2, 10/3)
# function delta(r::Float64) 8 * pdf(gd, r) + 0.2 end
    
function egpd_cdf(x::Float64, kappa::Float64, sigma::Float64, xi::Float64)
    sigma = max(sigma, 1e-12)
    kappa = max(kappa, 1e-12)

    if xi == 0.0
        return (1 - exp(-x/sigma))^kappa
    end

    base = 1 - (1 + xi*x/sigma)^(-1/xi)
    base = max(base, 0.0)

    return base^kappa
end


function egpd_logdens(x::Float64, kappa::Float64, sigma::Float64, xi::Float64)
    if xi == 0.0
        return log(kappa) - log(sigma) - x/sigma + (kappa-1)*log(1 - exp(-x/sigma))
    end
    t = 1 + xi*x/sigma
    t <= 0.0 && return -Inf
    log_gpd_dens = -log(sigma) + (-1/xi - 1)*log(t)
    F = max(1 - exp(-1/xi * log(t)), 1e-300)
    return log(kappa) + log_gpd_dens + (kappa-1)*log(F)
end


function dmegpd_biv(
    x1::Float64,
    x2::Float64,
    kappa::Float64,
    sigma::Float64,
    xi::Float64
)

    # Support is x1,x2 > 0
    if x1 <= 0.0 || x2 <= 0.0
        return 0.0
    end

    r = x1 + x2

    if r <= 0.0
        return 0.0
    end

    # Radial component
    log_term_rad = egpd_logdens(r, kappa, sigma, xi)

    if !isfinite(log_term_rad)
        return 0.0
    end

    # Angular scale
    delta_r = max(delta(r), 0.01)

    # log(x1/x2)
    log_ratio = log(x1) - log(x2)

    # Log Gaussian density
    log_term_ang =
        -0.5 * log(2π) -
        log(delta_r) -
        0.5 * (log_ratio / delta_r)^2 +
        log(r) -
        log(x1) -
        log(x2)

    return exp(log_term_rad + log_term_ang)
end

function get_cdf_val(
    target_x::Float64,
    x_prev::Float64,
    kappa::Float64,
    sigma::Float64,
    xi::Float64,
    norm_const::Float64,
)

    if target_x <= 0
        return 0.0
    end

    val, _ = quadgk(
        x -> dmegpd_biv(
            x,
            x_prev,
            kappa,
            sigma,
            xi
        ),
        0.0,
        target_x,
    )

    return val / norm_const
end

function sample_conditional(
    x_prev::Float64,
    kappa::Float64,
    sigma::Float64,
    xi::Float64,
)

    # Normalizing constant
    norm_const, _ = quadgk(
        x -> dmegpd_biv(
            x,
            x_prev,
            kappa,
            sigma,
            xi
        ),
        1e-10,
        Inf,
    )

    # Uniform draw
    p_target = rand()

    lower = 1e-10
    upper = max(10*x_prev, 1.0)

    f(x) = get_cdf_val(
        x,
        x_prev,
        kappa,
        sigma,
        xi,
        norm_const,
    ) - p_target

    # Expand search interval until it brackets the root
    while f(upper) < 0
        upper *= 2
    end

    return find_zero(f, (lower, upper), Bisection())
end

function simulate_megpd_chain(
    n_steps::Int,
    kappa::Float64,
    sigma::Float64,
    xi::Float64,
    x0::Float64=1.0,
    burn_in_prop::Float64=0.10,
    show_progress::Bool=true,
)

    0 <= burn_in_prop < 1 || throw(ArgumentError(
        "burn_in_prop must be in 0,1."
    ))

    x = Vector{Float64}(undef, n_steps)
    x[1] = x0

    if show_progress
        p = Progress(n_steps; desc="Simulating")
    end

    for t in 2:n_steps

        x[t] = sample_conditional(
            x[t-1],
            kappa,
            sigma,
            xi
        )

        if show_progress
            next!(p)
        end
    end

    burn_in = floor(Int, n_steps * burn_in_prop)

    return (
        full_chain=x,
        final_chain=x[(burn_in+1):end],
        burn_in=burn_in,
    )
end
